### PR TITLE
Sliding sync: Ignore invites from ignored users

### PR DESCRIPTION
**Pull request by MadLittleMods**
Ignore invites from ignored users (`m.ignored_user_list` in account data)

### Pull Request Checklist

* [x] Pull request is based on the develop branch
* [x] Pull request includes a [changelog file](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#changelog). The entry should:
  - Be a short description of your change which makes sense to users. "Fixed a bug that prevented receiving messages from other servers." instead of "Moved X method from `EventStore` to `EventWorkerStore`.".
  - Use markdown where necessary, mostly for `code blocks`.
  - End with either a period (.) or an exclamation mark (!).
  - Start with a capital letter.
  - Feel free to credit yourself, by adding a sentence "Contributed by @github_username." or "Contributed by [Your Name]." to the end of the entry.
* [x] [Code style](https://element-hq.github.io/synapse/latest/code_style.html) is correct
  (run the [linters](https://element-hq.github.io/synapse/latest/development/contributing_guide.html#run-the-linters))

### CODE REVIEW

File: synapse/handlers/sliding_sync/room_lists.py [hunk 1]
```diff
@@ -223,16 +223,30 @@ async def _compute_interested_rooms_new_tables(
...
         changes = await self._get_rewind_changes_to_current_membership_to_token(
             sync_config.user, room_membership_for_user_map, to_token=to_token
         )
         if changes:
-            room_membership_for_user_map = dict(room_membership_for_user_map)
             for room_id, change in changes.items():
                 if change is None:
                     # Remove rooms that the user joined after the `to_token`
-                    room_membership_for_user_map.pop(room_id)
+                    room_membership_for_user_map.pop(room_id, None)
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:249]: Unrelated but added a default to the `.pop(...)` so we don't run into a `KeyError` if the `room_id` is already removed from the dict.

File: synapse/handlers/sliding_sync/room_lists.py [hunk 2]
```diff
@@ -224,15 +224,31 @@ async def _compute_interested_rooms_new_tables(
             user_id
         )
 
+        # Remove invites from ignored users
+        ignored_users = await self.store.ignored_users(user_id)
+        if ignored_users:
+            # TODO: It would be nice to avoid these copies
+            room_membership_for_user_map = dict(room_membership_for_user_map)
```
Comment (MadLittleMods) [inline synapse/handlers/sliding_sync/room_lists.py:231]: Another reason to potentially add `SoftDeleteChainMap` (context https://github.com/element-hq/synapse/pull/17725#discussion_r1764394991)

Would tackle in a separate PR though.

File: tests/rest/client/sliding_sync/test_sliding_sync.py [hunk 3]
```diff
@@ -670,6 +671,116 @@ def test_forgotten_up_to_date(self) -> None:
...
+        # User1 ignores user2
+        channel = self.make_request(
+            "PUT",
+            f"/_matrix/client/v3/user/{user1_id}/account_data/{AccountDataTypes.IGNORED_USER_LIST}",
+            content={"ignored_users": {user2_id: {}}},
+            access_token=user1_tok,
+        )
+        self.assertEqual(channel.code, 200, channel.result)
+
+        # Sync once before we ignore to make sure the rooms can show up
```
Comment (devonh) [inline tests/rest/client/sliding_sync/test_sliding_sync.py:754]: This comment is incorrect. Otherwise the test looks good.

**Review by devonh — CHANGES_REQUESTED** [review]
Just a small change. Otherwise things look good.

File: tests/rest/client/sliding_sync/test_sliding_sync.py (hunk 3 above)
Comment (MadLittleMods) [reply tests/rest/client/sliding_sync/test_sliding_sync.py]: Updated :white_check_mark:

**Review by devonh — APPROVED** [review]

**Comment by MadLittleMods** [comment]
Thanks for the review @devonh 🐻‍❄️

**Pull request merged**